csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,290 @@
+      // inputs computed yet (tv_ids) and whether they can be computed
+      // (uncomputable_groups).
+      std::vector<ValGroup> uncomputed_producer_groups;
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        // NOTE: it suffices to use any Expr* in eg, since they are all
+        // guaranteed to have the same type and identical attributes.
+        Expr* expr = eg->front();
+        if (!uncomputed_producer_groups.empty()) {
+          // We already have something to compute in the next round, so skip.
+          continue;
+        }
+        std::vector<ValGroup> vg_inps;
+        std::vector<IterDomain*> id_inps;
+        bool all_inputs_computed = true;
+        for (Val* inp : expr->inputs()) {
+          ValGroup vg_inp = graph_->toGroup(inp);
+          vg_inps.push_back(vg_inp);
+          auto inp_it = tv_ids.find(vg_inp);
+          if (inp_it != tv_ids.end()) {
+            id_inps.push_back(inp_it->second);
+          } else {
+            // this input is not yet computed
+            all_inputs_computed = false;
+            if (uncomputable_groups.count(vg_inp) == 0) {
+              // Input is not yet proven to be incomputable, so try and
+              // compute it in the next iteration
+              uncomputed_producer_groups.push_back(vg_inp);
+            }
+          }
+        }
+        // some input is not yet computed
+        if (!all_inputs_computed) {
+          if (uncomputed_producer_groups.empty() && !id_inps.empty()) {
+            // There might be some uncomputable producer groups, but some are
+            // computed. Just pass those computed inputs through.
+            NVF_ERROR(id_inps.size() == 1);
+            tv_ids.emplace(vg, id_inps.front());
+          }
+          continue;
+        }
+        // Compute new ID expression
+        Expr* id_expr = nullptr;
+        if (auto* m = dynamic_cast<Merge*>(expr)) {
+          NVF_ERROR(id_inps.size() == 2);
+          auto* new_id = IterDomain::merge(id_inps[0], id_inps[1]);
+          id_expr = new_id->definition();
+        } else if (auto* s = dynamic_cast<Split*>(expr)) {
+          NVF_ERROR(id_inps.size() == 1);
+          auto new_ids =
+              IterDomain::split(id_inps[0], s->factor(), s->innerSplit());
+          id_expr = new_ids.first->definition();
+        } else if (auto* s = dynamic_cast<Swizzle*>(expr)) {
+          NVF_ERROR(id_inps.size() == 2);
+          auto new_ids =
+              IterDomain::swizzle(s->swizzleType(), id_inps[0], id_inps[1]);
+          id_expr = new_ids.first->definition();
+        } else if (auto* s = dynamic_cast<Swizzle2D*>(expr)) {
+          NVF_ERROR(id_inps.size() == 2);
+          auto new_ids = IterDomain::swizzle(
+              s->swizzleType(), id_inps[0], id_inps[1], s->swizzleMode());
+          id_expr = new_ids.first->definition();
+        } else {
+          NVF_ERROR(
+              false, "Unhandled IterDomain expression ", expr->toString());
+        }
```

Comment:
Can we reuse some of the existing replay APIs? For example, `ReplayTransform` in id_model/transform_replay.h?

Reply:
Thanks for the suggestion. I reused `ReplayTransformations` from `transform_iter.h` to avoid this tedious bit of code.